### ISSUE TITLE
Consider how to better handle CircuitBreaker failures

### ISSUE THREAD

**Issue by jhalterman** [jhalterman/failsafe#113]
Presently, execution failures according to a` CircuitBreaker` can cause state transitions in the CircuitBreaker, but the Failsafe event listeners are not called and a fallback is not used. Consider whether and how CircuitBreaker failures should be combined with event listeners and fallbacks, as they do with RetryPolicies. Are there any pitfalls here?

**Comment by dustpuppyNGTI** [comment]
I've been playing with the CircuitBreaker the last couple of days when I ran into this issue. I can't think of any pitfalls but I can say that the current behaviour of Failsafe made me want to file a bugreport.

To figure out the current behaviour I wrote a few tests:

`public class CircuitBreakerTests {

    private static final Integer INTERNAL_SERVER_ERROR = 500;

    @Test
    public void whenOpen() {
        CircuitBreaker breaker = new CircuitBreaker().withDelay(1, MINUTES);
        breaker.open();

        final boolean[] onSuccess = {false};
        final boolean[] onComplete = {false};
        final boolean[] onFailure = {false};
        final boolean[] fallback = {false};

        Failsafe.with(breaker)
            .onSuccess((o, executionContext) -> onSuccess[0] = true)
            .onFailure((result, failure, context) -> onFailure[0] = true )
            .onComplete((result, failure, context) -> onComplete[0] = true )
            .withFallback(() -> { fallback[0] = true; return true; })
            .get(() -> 200);

        assertFalse("Success SHOULD NOT be called", onSuccess[0]);
        assertTrue("Completion SHOULD be called", onComplete[0]);
        assertTrue("Success SHOULD be called", onFailure[0]);
        assertTrue("Fallback SHOULD be called", fallback[0]);
    }

    @Test
    public void onCodeWithPredicate() {
        final boolean[] onSuccess = {false};
        final boolean[] onComplete = {false};
        final boolean[] onFailure = {false};
        final boolean[] fallback = {false};

        CircuitBreaker breaker = new CircuitBreaker()
                .withDelay(1, MINUTES)
                .failIf((o, throwable) -> INTERNAL_SERVER_ERROR.equals(o) || throwable != null);

        Failsafe.with(breaker)
            .onSuccess((o, executionContext) -> onSuccess[0] = true)
            .onFailure((result, failure, context) -> onFailure[0] = true)
            .onComplete((result, failure, context) -> onComplete[0] = true)
            .withFallback(() -> { fallback[0] = true; return 200; })
            .get(() -> 500);

        assertFalse("Success SHOULD NOT be called", onSuccess[0]);
        assertTrue("Completion SHOULD be called", onComplete[0]);
        assertTrue("Success SHOULD be called", onFailure[0]);
        assertTrue("Fallback SHOULD be called", fallback[0]);
    }

    @Test
    public void onException() {
        final boolean[] onSuccess = {false};
        final boolean[] onComplete = {false};
        final boolean[] onFailure = {false};
        final boolean[] fallback = {false};

        CircuitBreaker breaker = new CircuitBreaker().withDelay(1, MINUTES);

        Failsafe.with(breaker)
                .onSuccess((o, executionContext) -> onSuccess[0] = true)
                .onFailure((result, failure, context) -> onFailure[0] = true)
                .onComplete((result, failure, context) -> onComplete[0] = true)
                .withFallback(() -> { fallback[0] = true; return 200; })
                .get(() -> { throw new Exception(); });

        assertFalse("Success SHOULD NOT be called", onSuccess[0]);
        assertTrue("Completion SHOULD be called", onComplete[0]);
        assertTrue("Success SHOULD be called", onFailure[0]);
        assertTrue("Fallback SHOULD be called", fallback[0]);
    }

    @Test
    public void onExceptionWithPredicate() {
        final boolean[] onSuccess = {false};
        final boolean[] onComplete = {false};
        final boolean[] onFailure = {false};
        final boolean[] fallback = {false};

        CircuitBreaker breaker = new CircuitBreaker()
            .withDelay(1, MINUTES)
            .failIf((o, throwable) -> INTERNAL_SERVER_ERROR.equals(o) || throwable != null);

        Failsafe.with(breaker)
            .onSuccess((o, executionContext) -> onSuccess[0] = true)
            .onFailure((result, failure, context) -> onFailure[0] = true)
            .onComplete((result, failure, context) -> onComplete[0] = true)
            .withFallback(() -> { fallback[0] = true; return 200; })
            .get(() -> { throw new Exception(); });

        assertFalse("Success SHOULD NOT be called", onSuccess[0]);
        assertTrue("Completion SHOULD be called", onComplete[0]);
        assertTrue("Success SHOULD be called", onFailure[0]);
        assertTrue("Fallback SHOULD be called", fallback[0]);
    }

    @Test
    public void onErrorWithWhen() {
        final boolean[] onSuccess = {false};
        final boolean[] onComplete = {false};
        final boolean[] onFailure = {false};
        final boolean[] fallback = {false};

        CircuitBreaker breaker = new CircuitBreaker()
            .withDelay(1, MINUTES)
            .failWhen(500);

        Failsafe.with(breaker)
            .onSuccess((o, executionContext) -> onSuccess[0] = true)
            .onFailure((result, failure, context) -> onFailure[0] = true)
            .onComplete((result, failure, context) -> onComplete[0] = true)
            .withFallback(() -> { fallback[0] = true; return 200; })
            .get(() -> 500);

        assertFalse("Success SHOULD NOT be called", onSuccess[0]);
        assertTrue("Completion SHOULD be called", onComplete[0]);
        assertTrue("Success SHOULD be called", onFailure[0]);
        assertTrue("Fallback SHOULD be called", fallback[0]);
    }
}`

Of these tests onException and onExceptionWithPredicate succeed. The statement "Presently, execution failures according to a CircuitBreaker can cause state transitions in the CircuitBreaker, but the Failsafe event listeners are not called and a fallback is not used." appears to me to be incorrect. If you configure a fallback and throw an Exception in the Callable FailSafe will call the fallback and the listeners.
However, in any other type of failure case the fallback and failure listeners will not be called.

I would expect this behaviour from Failsafe:
- the onComplete handler is always called
- the onSuccess handler is called only when the Runnable or Callable has been executed and did not generate an error condition
- the onFailure and fallback handlers are called when the CircuitBreaker is in OPEN state or when a predicate says it is a failure

**Comment by jhalterman** [comment]
Failsafe 2.0, which will be released soon, simplifies and clarifies this behavior a bit. `Fallback`, along with `RetryPolicy` and `CircuitBreaker`, are all `Policy` implementations, and may each have their own failure conditions. Each Policy makes its own determination about whether an execution result represents a failure. One Policy might consider something to be a failure and another not. Ex:

```java
Fallback fallback = new Fallback.of(this::connectToBackup).handle(ConnectException.class);
CircuitBreaker circuitBreaker = new CircuitBreaker().handle(Exception.class);

Failsafe.with(fallback, circuitBreaker).get(this::connect);
```

> I would expect this behaviour from Failsafe:
> the onComplete handler is always called
> the onSuccess handler is called only when the Runnable or Callable has been executed and did not generate an error condition
> the onFailure and fallback handlers are called when the CircuitBreaker is in OPEN state or when a predicate says it is a failure

`onComplete` will still be a top level event listener that is always called. `onSuccess` and `onFailure` can be registered on an individual Policy, or at the top level. Ex:

```java
RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
  .onSuccess(e -> log.info("retry policy succeeded"))
  .onFailure(e -> log.warn("retry policy failed"));

CircuitBreaker<Object> circuitBreaker = new CircuitBreaker<>()
  .onSuccess(e -> log.info("circuit breaker succeeded"))
  .onFailure(e -> log.warn("circuit breaker failed"));

Failsafe
  .with(retryPolicy, circuitBreaker)
  .onComplete(e -> log.info("done"))
  .onSuccess(e -> log.info("all policies were successful"))
  .onFailure(e -> log.warn("some policy failed"))
  .get(this::connect);
```

Hopefully this is more explicit and also straightforward. This will be released in 2.0.
